### ISSUE TITLE
CodeCache growth and lookup performance

### ISSUE THREAD

**Issue by olivergillespie** [jvm-profiling-tools/async-profiler#414]
I don't know much about what conditions cause new jmethodIDs to be generated, and existing ones to be invalidated, but I'm seeing some possibly dangerous behaviour in the CodeCache in my particular application as it grows and the lookups start to take a 'significant' amount of time. By significant, I mean my application's performance started to degrade to the point where there was always one thread with this stack: 

```
CodeCache::find(void const*)
Profiler::getAddressType(unsigned char*) Profiler::getJavaTraceAsync(void*, ASGCT_CallFrame*, int)
Profiler::recordSample(void*, unsigned long long, int, _jmethodID*, ThreadState) 
WallClock::signalHandler(int, siginfo_t*, void*)
```

so I guess the lookup started to take as much as the 100ms sampling interval. 

The jmethodID lookup appears to use a linear scan over the cached entries, so the lookup cost is proportional to the number of entries. When my application was behaving badly as above I attached `gdb` and I (hopefully correctly...) inspected the loop termination condition to infer a cache size of around 30 million, after 3 days of uptime. I recently added a log statement to monitor the cache size to confirm, so far it's steadily grown to a few hundred thousand entries within an hour or so but I'll see how it progresses.

Quirks of my usage: I'm running async-profiler in a semi-continous fashion, a la #71, for hours or days at a time in 5-minute 'chunks' (start, wait 5 minutes, stop, repeat). Not sure if this is at all a 'supported' use case. I'm running async-profiler 1.8.3 on OpenJDK 8u282, in wall mode with 100ms interval.

Since I don't know enough about the internals, I have no idea if this is likely a generic issue or if there's something specific to my application (overflow in the JVM code cache causing repeated compilation, perhaps?).

So basically I'd just like to share these observations from my usage in case there's anything useful for async-profiler to handle here.

**Comment by apangin** [comment]
Thank you for the report. I guess the described behavior may be observed when the JVM's Code Cache is almost full, and the JVM continuously evict and recompile methods.

When async-profiler expands its own CodeCache, it removes stale entries, however, the capacity increases indefinitely anyway. I'll fix this.

**Comment by olivergillespie** [comment]
Thanks!

If the JVM is repeatedly evicting and recompiling methods, what would that look like from the perspective of async-profiler's cache? Would the evicted entries be considered stale on the next cleanup? In my case it's the count which seems to be ever-growing, not just the capacity.

I'll try to check my JVM code cache behaviour too to see if that's really my issue.

**Comment by apangin** [comment]
I was able to reproduce the problem, and fixed that.
Please check if the fix helps in your case.

**Comment by olivergillespie** [comment]
Thank you. I'll try to test it when I get time. However I'm not sure it will tackle my specific case because my logging indicates that I don't have many entries being evicted.

```c++
void CodeCache::add(const void* start, int length, jmethodID method, bool update_bounds) {
    if (_count >= _capacity) {
        printf("Doubling CodeCache size, current capacity = %d\n", _capacity);
        expand();
        printf("After doubling CodeCache size, count = %d, capacity = %d\n", _count, _capacity);
    }
```

```
Doubling CodeCache size, current capacity = 256000
After doubling CodeCache size, count = 255674, capacity = 512000
Doubling CodeCache size, current capacity = 512000
After doubling CodeCache size, count = 511716, capacity = 1024000
Doubling CodeCache size, current capacity = 1024000
After doubling CodeCache size, count = 1023373, capacity = 2048000
```

I also checked my JVM code cache stats and it's stable at around 25% capacity.

I appreciate there's not much to go on so feel free to close this if you want, hopefully I'll get a chance to do some more debugging and find out what the new entries actually are.

**Comment by apangin** [comment]
Interesting. If `count` also increases, this is not the bug I was thinking of.

Can you please check

- whether these are Java compiled methods or runtime stubs (a simple way to check is to add debug printout in `Profiler::addJavaMethod` and `Profiler::addRuntimeStub`);
- JIT compiler stats: `jcmd <pid> PerfCounter.print | grep sun.ci`
- are there other agents loaded into JVM besides `libasyncProfiler`?

**Comment by olivergillespie** [comment]
I added debug logs, and the vast majority are from `Profiler::addJavaMethod`.

The application is also using the AspectJ weaver agent.

With the CodeCache entry count around 2,000,000, `PerfCounter.print` gives:

```
sun.ci.nmethodCodeSize=58576416
sun.ci.nmethodSize=111859392
sun.ci.osrBytes=192586
sun.ci.osrCompiles=303
sun.ci.osrTime=4402988334
sun.ci.standardBytes=6142847
sun.ci.standardCompiles=23527
sun.ci.standardTime=134576408967
sun.ci.threads=12
sun.ci.totalBailouts=12
sun.ci.totalCompiles=23830
sun.ci.totalInvalidates=0
```

I added logs to check exactly which methods were added from cache entry count of 128,000 to 256,000. This time period is considerably after the application seems to be basically stable with regards to compilation, so I figured it would be a good representation of the 'leaky' data.

I couldn't see any pattern in the newly added methods; they come from all parts of my application including my code, library code, JDK. Some analysis on the number of different methods and how often they were added to the CodeCache during this time shows there are a lot of different methods being added, and most of them are added multiple times.

**Number of times added to CodeCache**|**Count of methods**|**Total contribution to cache entry count**
:-----:|:-----:|:-----:
1|33|33
2|31|62
3|69|207
4|44|176
5|70|350
6|75|450
7|66|462
8|7086|56688
9|4869|43821
10|142|1420
11|42|462
12|13|156
13|3|39
14|5|70
15|6|90
16|400|6400
17|267|4539
18|244|4392
\>18|193|7596
TOTAL|13658|127413

Looking at the logs for a few random individual methods, I see a lot of them look like this:

```
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
Profiler::addJavaMethod(address=0x7f6d0295ae40, length=64, frameName=Ljava/lang/Long;compare)
```

Where they are repeatedly added with the same address and length, so it seems my cache is actually filling up with duplicates.

Since there's such a spike around frequency 8 and 9, I wonder if there's some global trigger re-emitting all of the compiled methods. I can look into what the AspectJ agent is up to, too.

**Comment by apangin** [comment]
Since there are only 23K compilations, I guess the only possibility for so many `CompiledMethodLoad` events is a repeated call to JVMTI `GenerateEvents` function.

Async-profiler calls `GenerateEvents` only once, so there must be some other agent calling this function.
There is a JDK bug [JDK-8222072](https://bugs.openjdk.java.net/browse/JDK-8222072) that results in the undesired interference between agents: when one agent calls `GenerateEvents`, the other one receives the events, too. The bug has been fixed in JDK 13+ and in recent JDK 11 update, but not in JDK 8.

Try adding `-XX:TraceJVMTI=CompiledMethodLoad+ts,GenerateEvents+ieo` JVM option. If my hypothesis is correct, the output will look like
```
JVMTI [Attach Listener] GenerateEvents {  event_type=68:CompiledMethodLoad
JVMTI [Attach Listener] method compile load event triggered
JVMTI [Attach Listener] class compile method load event sent java/lang/String.equals  
... many compile method load event messages ...
JVMTI [Attach Listener] GenerateEvents }
```
with a similar `GenerateEvents {` block repeated again and again.

Can it be the case you invoke async-profiler each time with a new .so library? E.g. unpack the profiler into a temporary folder, invoke it, cleanup the files, and so on? Check `/proc/<pid>/maps` to make sure the profiler is loaded just once.

**Comment by olivergillespie** [comment]
Ah!

Thanks a lot for those details. I added the JVMTI logs and `GenerateEvents` was invoked at regular intervals. Then I remembered that I'm also running https://github.com/jvm-profiling-tools/perf-map-agent in a loop, which is the 'other agent' you suspected. Specifically: https://github.com/jvm-profiling-tools/perf-map-agent/blob/master/src/c/perf-map-agent.c#L364 is the culprit.

Sorry it turned out to be not async-profiler's 'fault'. I guess the current CodeCache implementation works well in practice since entry count is generally low, but a couple of considerations which might be a useful safety net depending on the tradeoff:

- Don't store duplicate entries
- Improve lookup performance when count is large

Thanks for your help. Feel free to close if you don't want to make any other related changes.

**Comment by apangin** [comment]
I've committed the fix that

1. Warns if another agent calls `GenerateEvents()` on JDK 8;
2. Resets CodeCache before `GenerateEvents()` to prevent infinite growth of the method list.

I think there is no need to improve linear lookup, since it occurs only in rare cases when `AsyncGetCallTrace` fails.
